Clean stale left out buckets

shtripat opened the pull request:
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
If a bucket got removed while few nodes/disks were offline in cluster, while nodes/disks come back online, the stale bucket lies around as there is no bucket heal which gets triggered on its own. While scanning, if we find such stale buckets which dont have quorum anymore, safely we remove them now.

## Motivation and Context


## How to test this PR?
Step-1: Start a 4 node, 1 disk each MinIO setup as below
```
minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/

minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/ --address :9001

minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/ --address :9002

minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/ --address :9003
```

Step-2: Create two buckets `bkt1` and `bkt2` and load objects to them
```
$ mc alias set myminio http://127.0.0.1:9000 minioadmin minioadmin
$ mc mb myminio/bkt1
$ mc mb myminio/bkt2
$ mc cp /etc/hosts myminio/bkt2
```

Step-3: Now bring down the last MinIO instance and remove first bucket
`mc rb myminio/bkt1 --force`

Step-4: Verify that last disk only contains the stale bucket xl.meta
```
$ ll /tmp/bkt-test/1 
total 4
drwxr-xr-x. 1 shubhendu shubhendu 88 Oct 27 11:23 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .

$ ll /tmp/bkt-test/2 
total 4
drwxr-xr-x. 1 shubhendu shubhendu 88 Oct 27 11:23 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .

$ ll /tmp/bkt-test/3
total 4
drwxr-xr-x. 1 shubhendu shubhendu 88 Oct 27 11:23 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .

$ ll /tmp/bkt-test/4
total 4
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 36 Oct 27 11:24 .
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt1 <<<<<<<<========== HERE
drwxr-xr-x. 1 shubhendu shubhendu 72 Oct 27 11:24 .minio.sys
```

Step-5: Now start the last MinIO instance back and wait for few secs. Verify that last disk also dosn't contain the removed disk xl.meta anymore
```
$ ll /tmp/bkt-test/4 
total 4
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 72 Oct 27 11:25 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .
```

Without the fix, last disk retains the traces of removed bucket (while 4th MinIO instance was down). This fix as well will remove any dangling bucket which don't contain any data. If bucket contained data, it just logs as error once that dangling bucket cleanup didn't happen as bucket was not empty
```
While deleting dangling Bucket (bkt1), Drive /tmp/bkt-test/4 returned an error (volume is not empty) (*fmt.wrapError)
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

shtripat committed:
Clean stale left out buckets

If a bucket got removed while few nodes/disks were offline in
cluster, while nodes/disks come back online, the stale bucket
lies around as there is no bucket heal which gets triggered on
its own. While scanning, if we find such stale buckets which dont
have quorum anymore, safely we remove them now.

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

harshavardhana commented on cmd/peer-s3-client.go:
```diff
@@ -185,6 +185,14 @@ func (sys *S3PeerSys) ListBuckets(ctx context.Context, opts BucketOptions) ([]Bu
+		for bktName, count := range bucketsMap {
+			if count < quorum {
+				// Its safe to remove as its stale bucket lying around
+				sys.DeleteBucket(ctx, bktName, DeleteBucketOptions{Force: true})
```
quorum issue can happen where nodes are down, we cannot just assume that to be a precursor to delete a bucket entirely.

we should carefully look at the deleteDangling code for objects. We must be absolutely sure of what we are trying to do here.

This PR makes quorum-less a definitive way to purge the entire data automatically which would be unexpected.

shtripat replied:
sure, makes sense. let me explore more and redo this.

vadmeste replied:
I think let's try to avoid calling Deletebucket with Force set to true; It is okay if it fails to delete a bucket which is stale rather than always doing a force delete.

isObjectDangling(), this is a function that decides if an object is dangling or not; I think the logic is we only mark a bucket as dangling in well defined use cases; meaning if we do not have certain errors and secnarios, we should not dare to remove a bucket.

shtripat committed:
Incorporated review comments

Added a function `isBucketDangling` and used for deciding
whether to remove the bucket.

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

vadmeste commented on cmd/erasure-healing.go:
```diff
@@ -1019,6 +1032,22 @@ func isAllNotFound(errs []error) bool {
+	}
+	notFoundCount := 0
+	for _, err := range errs {
+		if err != nil && (err.Error() == errFileNotFound.Error() || err.Error() == errFileCorrupt.Error()) {
```
So isAllBucketsNotFound() takes errs from readAllBucketInfo();

readAllBucketInfo() runs xl-storage.StatVol(), which returns errVolumeNotFound, not errFileNotFound; 

errFileCorrupt is not returned as well, so no need to check for it

shtripat replied:
ah yes, that's a mistake. Will change to use `errVolumeNotFound` and remove `errFileCorrupt`

vadmeste commented on cmd/erasure-object.go:
```diff
@@ -545,6 +559,46 @@ func (er erasureObjects) deleteIfDangling(ctx context.Context, bucket, object st
+				err := disks[index].DeleteVol(ctx, bucket, false)
+				if errors.Is(err, errVolumeNotEmpty) {
+					// If volume non empty, force delete as its dangling anyway
+					return disks[index].DeleteVol(ctx, bucket, true)
```
I am really not in favor of doing Force Delete here; the benefit is low comparing to the risk

shtripat replied:
ok, so no need to have force delete called at all? will remove.

shtripat replied:
As scanner process wont consider a dangling bucket for clearing up its dangling objects, and we are sure that volume metadata is found on more than N/2 nodes of the cluster, isn't it safe to remove bucket with force after first error with `errVolumeNotEmpty`?
@vadmeste @harshavardhana please suggest.

harshavardhana replied:
My suggestion is @shtripat let's leave out the force part instead just perform normal DelVol if possible if its not they engage support and we investigate.

Keeping it manual is safer.

shtripat replied:
ok, I will remove the force delete part here, but in actual dangling bucket scenario where objects as well left out, this code wont have any effect as bucket wont be deleted due to non empty. We can just log an error once that during removal of dangling buckets, it was not able to delete as bucket was non empty.

vadmeste commented on cmd/erasure-healing.go:
```diff
@@ -1133,6 +1162,35 @@ func isObjectDangling(metaArr []FileInfo, errs []error, dataErrs []error) (valid
+			switch {
+			case errors.Is(readErr, errVolumeNotFound):
+				notFoundCount++
+			case errors.Is(readErr, errFileCorrupt):
```
No file corrupt error in my opinion is returned in this case

shtripat replied:
ack. will remove

vadmeste commented on cmd/erasure-healing.go:
```diff
@@ -1133,6 +1162,35 @@ func isObjectDangling(metaArr []FileInfo, errs []error, dataErrs []error) (valid
+		var (
+			notFoundCount     int
+			corruptedCount    int
+			diskNotFoundCount int
```
This doesn't seem to be needed, you can remove it

shtripat replied:
ack

shtripat replied:
actually we are concerned about volume not found error `errVolumeNotFound` so we can calculate that count and use. will change accordingly

vadmeste reviewed (commented):
Few comments

shtripat committed:
Log error once if removal of dangling bucket fails

This failure can happen due to bucket not being empty. In such
case we wont remove the bucket explicitly and rather user needs
to take cautious call to manually remove the dangling bucket from
the disk where it lies.

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Corrected check for volume not found.

Also just log the error once if unable to delete dangling bucket

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

harshavardhana commented on cmd/peer-s3-client.go:
```diff
@@ -185,6 +186,21 @@ func (sys *S3PeerSys) ListBuckets(ctx context.Context, opts BucketOptions) ([]Bu
+			if count < quorum {
+				// Queue a bucket heal task
+				globalHealStateLK.Lock()
+				bgSeq, ok := globalBackgroundHealState.getHealSequenceByToken(bgHealingUUID)
```
We should put this to MRF not global heal queue @shtripat 

A much simpler code to handle it's called er.addPartial()

shtripat replied:
Done

shtripat committed:
Put to MRF for healing bucket

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

harshavardhana commented on cmd/erasure-healing.go:
```diff
@@ -1133,6 +1161,35 @@ func isObjectDangling(metaArr []FileInfo, errs []error, dataErrs []error) (valid
+// Bucket is considered dangling/corrupted if and only
+// if bucket meta data is missing on more than N/2 disks
+func isBucketDangling(metaArr []VolInfo, errs []error) bool {
+	// We can consider a bucket not reliable
+	// when xl.meta is not found in read quorum disks.
+	// or when xl.meta is not readable in read quorum disks.
```
these comments don't make sense for buckets

shtripat replied:
Will remove

harshavardhana commented on cmd/erasure-healing.go:
```diff
@@ -1133,6 +1161,35 @@ func isObjectDangling(metaArr []FileInfo, errs []error, dataErrs []error) (valid
+				diskNotFoundCount++
+			}
+		}
+		return notFoundCount, corruptedCount, diskNotFoundCount
```
```suggestion
		return notFoundCount, corruptedCount, diskNotFoundCount
```

diskNotFoundCount is not used? if there is any drive that is offline we shouldn't treat it as dangling and return right away.

shtripat replied:
Ack

harshavardhana commented on cmd/erasure-healing.go:
```diff
@@ -1133,6 +1161,35 @@ func isObjectDangling(metaArr []FileInfo, errs []error, dataErrs []error) (valid
+		return notFoundCount, corruptedCount, diskNotFoundCount
+	}
+
+	notFoundMetaErrs, corruptedMetaErrs, _ := danglingErrsCount(errs)
```
```suggestion
	notFoundMetaErrs, corruptedMetaErrs, diskNotFoundErrs := danglingErrsCount(errs)
	if diskNotFoundErrs > 0 {
	    return false
	}
```

shtripat replied:
Ack

harshavardhana reviewed (changes requested):
Unfortunately looks like we would have to re-implement this as part of the `peer` call such that dangling buckets are cleared based on the per pool quorum.

Look at the implementation at `s3-peer-client.go`
